server/app/services/program/ActiveAndDraftPrograms.java

```diff
@@ -70,30 +81,104 @@ private ImmutableMap<String, ProgramDefinition> mapNameToProgram(
     // Note: Building this lookup has N+1 query behavior since a call to getProgramDefinition does
     // an additional database lookup in order to sync the set of questions associated with the
     // program.
+
+    // Active, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> activeNameToProgram =
+        mapNameToProgramWithFilter(repository, service, active, Optional.of(DisplayMode.DISABLED));
+
+    // All active programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> activeNameToProgramAll =
         mapNameToProgram(repository, service, active);
 
+    // Draft, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> draftNameToProgram =
+        mapNameToProgramWithFilter(repository, service, draft, Optional.of(DisplayMode.DISABLED));
+
+    // All draft programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> draftNameToProgramAll =
         mapNameToProgram(repository, service, draft);
 
-    this.activePrograms = activeNameToProgram.values().asList();
-    this.draftPrograms = draftNameToProgram.values().asList();
-    this.versionedByName =
-        Sets.union(activeNameToProgram.keySet(), draftNameToProgram.keySet()).stream()
-            .collect(
-                ImmutableMap.toImmutableMap(
-                    Function.identity(),
-                    programName -> {
-                      return Pair.create(
-                          Optional.ofNullable(activeNameToProgram.get(programName)),
-                          Optional.ofNullable(draftNameToProgram.get(programName)));
-                    }));
+    this.activePrograms =
+        types.equals(allProgramTypes)
+            ? activeNameToProgramAll.values().asList()
+            : activeNameToProgram.values().asList();
+    this.draftPrograms =
+        types.equals(allProgramTypes)
+            ? draftNameToProgramAll.values().asList()
+            : draftNameToProgram.values().asList();
+    if (types.equals(allProgramTypes)) {
+      this.versionedByName =
+          createVersionedByNameMap(activeNameToProgramAll, draftNameToProgramAll);
+    } else if (types.equals(EnumSet.of(ActiveAndDraftProgramsType.DISABLED))) {
+      // Disabled active programs.
+      ImmutableMap<String, ProgramDefinition> disabledActiveNameToProgram =
+          filterMapNameToProgram(activeNameToProgramAll, activeNameToProgram);
+      // Disabled draft programs.
+      ImmutableMap<String, ProgramDefinition> disabledDraftNameToProgram =
+          filterMapNameToProgram(draftNameToProgramAll, draftNameToProgram);
+      this.versionedByName =
+          createVersionedByNameMap(disabledActiveNameToProgram, disabledDraftNameToProgram);
+    } else if (types.equals(EnumSet.of(ActiveAndDraftProgramsType.IN_USE))) {
+      this.versionedByName = createVersionedByNameMap(activeNameToProgram, draftNameToProgram);
+    } else {
+      throw new IllegalArgumentException("Unsupported ActiveAndDraftProgramsType: " + types);
```

Comment:
**suggestion:** I don't think we should need this with the new implementation

Reply:
We do need this even with the new implementation, otherwise it errors `variable versionedByName might not have been initialized`. Feel free to run the branch locally.